jwst/associations/lib/rules_level2_base.py

```diff
@@ -1060,18 +968,181 @@ def _init_hook(self, item):
     nodding between different slits, or internal slit positions.
     The main associations rules will collect all the exposures of all the nods
     into a single association. Then, on finalization, this one association
     is split out into many associations, where each nod is, in turn, treated
     as science, and the other nods are treated as background.
     """
 
+    @staticmethod
+    def nod_background_overlap(science_item, background_item):
+        """
+        Check that a candidate background nod will not overlap with science.
+
+        For NIRSpec fixed slit or MOS data, this returns True if the
+        background candidate shares a primary dither point with the
+        science.
+
+        In addition, for NIRSpec fixed slit exposures taken with slit
+        S1600A1 in a 5-point nod pattern, this returns True when the
+        background candidate is in the next closest primary position
+        to the science.
+
+        For any other data, this function returns False (no overlap).
+        """
+        # Get exp_type, needed for any data:
+        # if not present, return False
+        try:
+            exptype = str(science_item['exp_type']).lower()
+        except KeyError:
+            return False
+
+        # Return False for any non-FS or MOS data
+        if exptype not in ['nrs_fixedslit', 'nrs_msaspec']:
+            return False
+
+        # Get pattern number values, needed for FS or MOS
+        try:
+            numdthpt = int(science_item['numdthpt'])
+            patt_num = int(science_item['patt_num'])
+            bkg_num = int(background_item['patt_num'])
+        except (KeyError, ValueError):
+            numdthpt = 0
+            patt_num = None
+            bkg_num = None
+
+        # Get subpxpts (or old alternate 'subpxpns')
+        try:
+            subpx = int(science_item['subpxpts'])
+        except (KeyError, ValueError):
+            try:
+                subpx = int(science_item['subpxpns'])
+            except (KeyError, ValueError):
+                subpx = None
+        try:
+            bkg_subpx = int(background_item['subpxpts'])
+        except (KeyError, ValueError):
+            try:
+                bkg_subpx = int(background_item['subpxpns'])
+            except (KeyError, ValueError):
+                bkg_subpx = None
+
+        # If values not found, return False for MOS, True for FS
+        if (None in [patt_num, bkg_num, subpx, bkg_subpx]
+                or subpx == 0 or bkg_subpx == 0):
+            if exptype == 'nrs_fixedslit':
+                # These values are required for background
+                # candidates for FS: report an overlap.
+                return True
+            else:
+                # For MOS, it's okay to go ahead and use them.
+                # Report no overlap.
+                return False
+
+        # Check for primary point overlap
+        sci_prim_dithpt = (patt_num - 1) // subpx
```

Comment:
Ooh, good catch, thank you!  Do you have test data you saw this with?

Reply:
nope, I don't know if we'll run into that case with a dataset but figured it's better to have it covered, just in case

Reply:
Okay, I'm now looking at the code, and I think there's not actually a possibility of raising a TypeError there.  In line 1033, if there is a None in any of those values, it will either return True or False, depending on the exposure type.  It never gets to line 1046 in that condition.